File: banco_problemas_aplicados/nivel_intermedio/soluciones/java/Problema30.java
/*
 * Problema 30 - Transposicion de Matriz de Admitancias
 * Devuelve la transpuesta M x N:  Trans[j][i] = Mat[i][j].
 */
import java.util.Scanner;
import java.util.Locale;

public class Problema30 {

    static double[][] transpone(double[][] mat) {
        int n = mat.length;
        int m = mat[0].length;
        double[][] trans = new double[m][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                trans[j][i] = mat[i][j];
            }
        }
        return trans;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Numero de filas (N): ");
        int n = sc.nextInt();
        System.out.print("Numero de columnas (M): ");
        int m = sc.nextInt();

        double[][] mat = new double[n][m];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                System.out.print("mat[" + i + "][" + j + "]: ");
                mat[i][j] = sc.nextDouble();
            }
        }

        double[][] trans = transpone(mat);

        System.out.println("Matriz transpuesta (" + m + " x " + n + "):");
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                System.out.printf(Locale.US, "%.2f ", trans[i][j]);
            }
            System.out.println();
        }
        sc.close();
    }
}

File: banco_problemas_aplicados/nivel_intermedio/soluciones/c/problema30.c
/*
 * Problema 30 - Transposicion de Matriz de Admitancias
 * Contexto: analisis matricial de circuitos. Recibe una matriz N x M y devuelve su
 * transpuesta M x N:  Trans[j][i] = Mat[i][j].
 */
#include <stdio.h>

void transpone(int n, int m, double mat[n][m], double trans[m][n]) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            trans[j][i] = mat[i][j];    /* se intercambian los indices */
        }
    }
}

int main(void) {
    int n, m;
    printf("Numero de filas (N): ");
    scanf("%d", &n);
    printf("Numero de columnas (M): ");
    scanf("%d", &m);

    double mat[n][m];
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            printf("mat[%d][%d]: ", i, j);
            scanf("%lf", &mat[i][j]);
        }
    }

    double trans[m][n];
    transpone(n, m, mat, trans);

    printf("Matriz transpuesta (%d x %d):\n", m, n);
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            printf("%.2f ", trans[i][j]);
        }
        printf("\n");
    }
    return 0;
}
